go/storage/mkvs/db/badger/migrate.go

```diff
@@ -665,15 +681,647 @@ func (v4 *v4Migrator) Migrate() (rversion uint64, rerr error) {
+		default:
+			return false, fmt.Errorf("error reading updated nodes metadata: %w", err)
+		}
+
+		v5.helper.DisplayStep(fmt.Sprintf("migrated root %s -> %s", root, newRoot))
+	}
+
+	// Update roots metadata.
+	roots.Roots = newRoots
+	entry := badger.NewEntry(v5RootsMetadataKeyFmt.Encode(version), cbor.Marshal(roots))
+	err = v5.changeBatch.SetEntryAt(entry, tsMetadata)
+	if err != nil {
+		return false, fmt.Errorf("error setting updated roots metadata key: %w", err)
+	}
+
+	// Save progress.
+	v5.meta.LastMigratedRoots = migratedRoots
+	v5.meta.LastMigratedVersion = &version
+
+	if err = v5.flush(false); err != nil {
+		return false, err
+	}
+
+	// Delete any pending keys.
+	for _, k := range v5.deleteKeys {
+		if err = v5.changeBatch.DeleteAt(k.key, k.ts); err != nil {
+			return false, fmt.Errorf("can't delete old key %x: %w", k.key, err)
+		}
+	}
+	v5.deleteKeys = nil
+
+	return true, nil
+}
+
+func (v5 *v5Migrator) maybePruneNode(h hash.Hash, version uint64) error {
+	item, err := v5.readTxn.Get(v4NodeKeyFmt.Encode(&h))
+	if err != nil {
+		return fmt.Errorf("error loading node %s: %w", h, err)
+	}
+	raw, err := item.ValueCopy(nil)
+	if err != nil {
+		return fmt.Errorf("error loading node %s: %w", h, err)
+	}
+
+	// Decode node type.
+	switch raw[0] {
+	case node.PrefixInternalNode:
+		var n node.InternalNode
+		_, err = v5.unmarshalV4InternalNode(&n, raw)
+		if err != nil {
+			return fmt.Errorf("error decoding internal node %s: %w", h, err)
+		}
+
+		// Maybe prune children.
+		for _, child := range []*node.Pointer{n.Left, n.Right} {
+			if child == nil {
+				continue
+			}
+			err = v5.maybePruneNode(child.Hash, version)
+			if err != nil {
+				return err
+			}
+		}
+	case node.PrefixLeafNode:
+	default:
+		return fmt.Errorf("unknown node type for node %s: 0x%02x", h, raw[0])
+	}
+
+	if tsToVersion(item.Version()) == version {
+		if err = v5.changeBatch.DeleteAt(item.KeyCopy(nil), item.Version()); err != nil {
+			return fmt.Errorf("error pruning node %s: %w", h, err)
+		}
+	}
+	return nil
+}
+
+func (v5 *v5Migrator) pruneVersion(version uint64) error {
+	defer func() {
+		v5.readTxn.Discard()
+		v5.readTxn = v5.db.db.NewTransactionAt(maxTimestamp, false)
+	}()
+
+	// Remove all roots in version.
+	var rootsMeta v4RootsMetadata
+	item, err := v5.readTxn.Get(v4RootsMetadataKeyFmt.Encode(version))
+	switch err {
+	case nil:
+	case badger.ErrKeyNotFound:
+		// Version does not exist.
+		return nil
+	default:
+		return fmt.Errorf("error reading roots metadata for version %d: %w", version, err)
+	}
+	err = item.Value(func(data []byte) error {
+		return cbor.UnmarshalTrusted(data, &rootsMeta)
+	})
+	if err != nil {
+		return fmt.Errorf("error decoding roots metadata for version %d: %w", version, err)
+	}
+
+	maybeLoneRoots := make(map[typedHash]bool)
+	for rootHash, derivedRoots := range rootsMeta.Roots {
+		if len(derivedRoots) == 0 {
+			// Need to only set the flag iff the flag has not already been set
```

Comment:
`rootsMeta.Roots` is `map[typedHash][]typedHash` right? 
So each key (`rootHash`) will only appear once so this check is unnecessary?

Reply:
Hm that's a good point, so I copied this from the existing `Prune` implementation where it also doesn't seem to be necessary. Based on blame it seems that this originated from the time when there was no `Roots` map and this was inside a loop which iterated over Badger keys so uniqueness was not guaranteed. Will remove it in both places.

Reply:
Fixed.

Reply:
Cool thanks!